test/realtime/message.test.js

```diff
@@ -1183,5 +1183,135 @@ define(['ably', 'shared_helper', 'async', 'chai'], function (Ably, helper, async
+              number: 12345,
+              bool: false,
+            },
+          },
+        },
+        {
+          name: 'end',
+          data: 'last message check',
+        },
+      ];
+
+      var filterOption = {
+        filter: 'name == `"filtered"` && headers.number == `26095`',
+      };
+
+      try {
+        /* set up realtime */
+        var realtime = helper.AblyRealtime({ key: helper.getTestApp().keys[5].keyStr });
+        var rest = helper.AblyRest();
+
+        realtime.connection.on('connected', function () {
+          var rtFilteredChannel = realtime.channels.getDerived('chan', filterOption);
+          var rtUnfilteredChannel = realtime.channels.get('chan');
+
+          var filteredMessages = [];
+          var unFilteredMessages = [];
+          /* subscribe to event */
+          rtFilteredChannel.subscribe(function (msg) {
+            try {
+              // Push received filtered messages into an array
+              filteredMessages.push(msg);
+            } catch (err) {
+              closeAndFinish(done, realtime, err);
+              return;
+            }
+          });
+
+          rtUnfilteredChannel.subscribe(function (msg) {
+            try {
+              // Push received unfiltered messages into an array
+              unFilteredMessages.push(msg);
+            } catch (err) {
+              closeAndFinish(done, realtime, err);
+              return;
+            }
+          });
+
+          // Subscription to check all messages were received as expected
+          rtUnfilteredChannel.subscribe('end', function (msg) {
+            try {
+              expect(msg.data).to.equal(testData[testData.length - 1].data, 'Unexpected msg data received');
+
+              // Check that we receive expected messages on filtered channel
+              expect(filteredMessages.length).to.equal(2, 'Expect only two filtered message to be received');
+              expect(filteredMessages[0].data).to.equal(testData[0].data, 'Unexpected data received');
+              expect(filteredMessages[1].data).to.equal(testData[2].data, 'Unexpected data received');
+              expect(filteredMessages[0].extras.headers.name).to.equal(
+                testData[0].extras.headers.name,
+                'Unexpected header value received'
+              );
+              expect(filteredMessages[1].extras.headers.name).to.equal(
+                testData[2].extras.headers.name,
+                'Unexpected header value received'
+              );
+              // Check that message with header that doesn't meet filtering condition is not received.
+              for (msg of filteredMessages) {
+                expect(msg.extras.headers.number).to.equal(26095, 'Unexpected header filtering value received');
+              }
+
+              // Check that we receive expected messages on unfiltered channel
+              expect(unFilteredMessages.length).to.equal(6, 'Expect only 6 unfiltered message to be received');
```

Comment:
possibly worth adding a comment to say that the extra message here is the `end` message